From 9be36fb8cb5bcdab5bbbc7a36388dc9a3d1cc397 Mon Sep 17 00:00:00 2001
From: Joseph Myers <joseph@codesourcery.com>
Date: Wed, 26 Mar 2014 18:59:08 +0000
Subject: [PATCH] Make x86_64 fegetenv preserve exception mask (bug 16198).

Bug 16198 is x86_64 fegetenv wrongly masking exceptions for which
traps are enabled, because that's a side-effect of the fnstenv
instruction.  This patch fixes it to use fldenv immediately after
fnstenv, like the i386 version.  Tested x86_64 and x86.

	[BZ #16198]
	* sysdeps/x86_64/fpu/fegetenv.c (fegetenv): Use fldenv after
	fnstenv.
	* math/test-fenv-preserve.c: New file.
	* math/Makefile (tests): Add test-fenv-preserve.
---
 math/Makefile                 |  2 +-
 math/test-fenv-preserve.c     | 53 +++++++++++++++++++++++++++++++++++++++++++
 sysdeps/x86_64/fpu/fegetenv.c |  3 +++
 5 files changed, 69 insertions(+), 5 deletions(-)
 create mode 100644 math/test-fenv-preserve.c

diff --git a/math/Makefile b/math/Makefile
index 09d14a0..c13ca80 100644
--- a/math/Makefile
+++ b/math/Makefile
@@ -90,7 +90,7 @@ tests = test-matherr test-fenv atest-exp atest-sincos atest-exp2 basic-test \
 	test-misc test-fpucw test-fpucw-ieee tst-definitions test-tgmath \
 	test-tgmath-ret bug-nextafter bug-nexttoward bug-tgmath1 \
 	test-tgmath-int test-tgmath2 test-powl tst-CMPLX tst-CMPLX2 test-snan \
-	test-fenv-tls $(tests-static)
+	test-fenv-tls test-fenv-preserve $(tests-static)
 tests-static = test-fpucw-static test-fpucw-ieee-static
 # We do the `long double' tests only if this data type is available and
 # distinct from `double'.
diff --git a/math/test-fenv-preserve.c b/math/test-fenv-preserve.c
new file mode 100644
index 0000000..8288b2c
--- /dev/null
+++ b/math/test-fenv-preserve.c
@@ -0,0 +1,53 @@
+/* Test fegetenv preserves exception mask (bug 16198).
+   Copyright (C) 2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <fenv.h>
+#include <stdio.h>
+
+static int
+do_test (void)
+{
+  fenv_t env;
+
+  if (feenableexcept (FE_INVALID) != 0)
+    {
+      puts ("feenableexcept (FE_INVALID) failed, cannot test");
+      return 0;
+    }
+
+  if (fegetenv (&env) != 0)
+    {
+      puts ("fegetenv failed, cannot test");
+      return 0;
+    }
+
+  int ret = fegetexcept ();
+  if (ret == FE_INVALID)
+    {
+      puts ("fegetenv preserved exception mask, OK");
+      return 0;
+    }
+  else
+    {
+      printf ("fegetexcept returned %d, expected %d\n", ret, FE_INVALID);
+      return 1;
+    }
+}
+
+#define TEST_FUNCTION do_test ()
+#include "../test-skeleton.c"
diff --git a/sysdeps/x86_64/fpu/fegetenv.c b/sysdeps/x86_64/fpu/fegetenv.c
index 364f917..0e0c269 100644
--- a/sysdeps/x86_64/fpu/fegetenv.c
+++ b/sysdeps/x86_64/fpu/fegetenv.c
@@ -22,6 +22,9 @@ int
 fegetenv (fenv_t *envp)
 {
   __asm__ ("fnstenv %0\n"
+	   /* fnstenv changes the exception mask, so load back the
+	      stored environment.  */
+	   "fldenv %0\n"
 	   "stmxcsr %1" : "=m" (*envp), "=m" (envp->__mxcsr));
 
   /* Success.  */
-- 
1.9.0

